pkg/operator/nsfinalizercontroller/finalizer_controller.go

```diff
@@ -0,0 +1,181 @@
+package nsfinalizercontroller
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubeinformers "k8s.io/client-go/informers"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	appsv1lister "k8s.io/client-go/listers/apps/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+type finalizerController struct {
+	name          string
+	namespaceName string
+
+	namespaceGetter v1.NamespacesGetter
+	podLister       corev1listers.PodLister
+	dsLister        appsv1lister.DaemonSetLister
+	eventRecorder   events.Recorder
+
+	preRunHasSynced []cache.InformerSynced
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+// NewFinalizerController is here because
+// When running an aggregated API on the platform, you delete the namespace hosting the aggregated API. Doing that the
+// namespace controller starts by doing complete discovery and then deleting all objects, but pods have a grace period,
+// so it deletes the rest and requeues. The ns controller starts again and does a complete discovery and.... fails. The
+// failure means it refuses to complete the cleanup. Now, we don't actually want to delete the resoruces from our
+// aggregated API, only the server plus config if we remove the apiservices to unstick it, GC will start cleaning
+// everything. For now, we can unbork 4.0, but clearing the finalizer after the pod and daemonset we created are gone.
```

Comment:
with so much apiserver specifics, why not put everything into `pkg/operator/apiserver/controller/namespacefinalizer`, and the others in similar places?